solutions/gold/izho-17-LongestBeautifulSequence.mdx

```diff
@@ -0,0 +1,270 @@
+
+More specifically, we redefine $\texttt{dp}[i]$ to be the length of the longest LBS that ends with **value** $i$ (rather than index). To calculate $\texttt{dp}[a_i]$, we try to transition from $\texttt{dp}[j]$ for all $j$ that satisfy $bc[j][a_i] == k[i]$. This improves our time complexity to $\mathcal{O}(NM)$.
+
+### Implementation
+
+**Time Complexity**: $\mathcal{O}(NM)$
+
+<Spoiler title="Code">
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+#define f first
+#define s second
+using namespace std;
+using pii = pair<int, int>;
+const int B = 8;
+
+int main() {
+	// preprocess:
+	// bc[i][j] = bit_count(i & j)
+	vector<vector<int>> bc(1 << B, vector<int>(1 << B));
+	for (int i = 0; i < (1 << B); i++) {
+		for (int j = 0; j < (1 << B); j++) {
+			bc[i][j] = __builtin_popcount(i & j);
+		}
+	}
+
+	int n;
+	cin >> n;
+	vector<int> a(n);
+	for (int i = 0; i < n; i++) cin >> a[i];
+	vector<int> k(n);
+	for (int i = 0; i < n; i++) cin >> k[i];
+	int ans = 1, best_i = 0;  // best_i: best index to end subsequence at
+	vector<int> prv(n);       // prv[i]: best index to include before i
+	iota(prv.begin(), prv.end(), 0);  // prv[i] = i
+	vector<pii> dp(1 << B, make_pair(-1, -1));
+	// dp[i].f: longest lbs that ends in i
+	// dp[i].s: optimal end index for lbs that ends in i
+
+	for (int i = 0; i < n; i++) {
+		if (dp[a[i]].f == -1) {
+			dp[a[i]].f = 1;
+			dp[a[i]].s = i;
+		}
+		for (int j = 0; j < (1 << B); j++) {
+			if (bc[j][a[i]] == k[i] && dp[j].f + 1 > dp[a[i]].f) {
+				dp[a[i]].f = dp[j].f + 1;
+				dp[a[i]].s = i;
+				prv[i] = dp[j].s;
+			}
+		}
+		if (dp[a[i]].f > ans) {
+			ans = dp[a[i]].f;
+			best_i = i;
+		}
+	}
+	cout << ans << endl;
+	vector<int> res;
+	while (prv[best_i] != best_i) {
+		res.push_back(best_i);
+		best_i = prv[best_i];
+	}
+	res.push_back(best_i);
+	reverse(res.begin(), res.end());
+	for (int x : res) cout << x + 1 << " ";
+}
+```
+
+</CPPSection>
+
+</LanguageSection>
+</Spoiler>
+
+### Subtask 4 (Full Credit)
+
+With much higher bounds on both $N$ and $M$, neither of the previously presented solutions will run in time.
+
+In our solution for Subtask 3, the loop from $0...M$ seems repetitive and motivates somehow restricting the search space of DP states. The most extreme way to achieve this is to embed all search parameters directly in the DP state (i.e. $\texttt{dp}[i][k]$ is the length of the longest LBS that ends in $x$, where $\texttt{bc}[x][i] == k$). $\texttt{dp}[a[i]][k[i]]$ is now an $\mathcal{O}(1)$ transition to find the longest LBS that ends at index $i$. However, each DP state now encapsulates a much larger amount of possible LBS's: any LBS that ends in $x$, where $\texttt{bc}[x][i] == k$, will be factored into $\texttt{dp}[i][k]$. Similarly, an LBS with fixed final value $a_i$ now affects $M$, rather than $1$, DP states: namely, all states $\texttt{dp}[x][k]$ where $x \in [0, M)$ and $k = \texttt{bc}[a_i][x]$.
```

Comment:
10 of my brain cells died trying to understand this
can you break it up?

Reply:
tried but it's probably still a bit unclear; fresh pair of eyes for suggestions would be greatly appreciated